pkg/backup/backupsink/sst_sink_key_writer.go

```diff
@@ -0,0 +1,310 @@
+	lastFile := &s.flushedFiles[len(s.flushedFiles)-1]
+	if newSpan.Equal(lastFile.Span) {
+		return nil
+	}
+	samePrefix, err := sameElidedPrefix(newSpan.Key, lastFile.Span.EndKey, s.conf.ElideMode)
+	if err != nil {
+		return err
+	}
+	isOverlapping := newSpan.Key.Compare(lastFile.Span.EndKey) < 0
+	if !samePrefix || isOverlapping {
+		if isOverlapping {
+			log.VEventf(
+				ctx, 1, "flushing backup file %s of size %d because span %v overlaps with %v",
+				s.outName, s.flushedSize, newSpan, lastFile.Span,
+			)
+		} else {
+			log.VEventf(
+				ctx, 1,
+				"flushing backup file %s of size %d because span %v has a different prefix than %v",
+				s.outName, s.flushedSize, newSpan, lastFile.Span,
+			)
+		}
+		s.stats.oooFlushes++
+		if err := s.Flush(s.ctx); err != nil {
+			return err
+		}
+		return nil
+	}
+	return nil
+}
+
+// maybeFlushOOOKey checks if the new key is out of order with respect to the
+// last key written using WriteKey to the SST file or if they do not share the
+// same prefix based on the sink's ElideMode. If either of these are true, the
+// file is flushed and a new one is created.
+func (s *SSTSinkKeyWriter) maybeFlushOOOKey(ctx context.Context, newKey roachpb.Key) error {
+	if s.prevKey == nil {
+		return nil
+	}
+	if len(s.flushedFiles) == 0 {
+		panic("unexpected empty flushedFiles despite prevKey being set")
+	}
+	lastSpan := s.flushedFiles[len(s.flushedFiles)-1].Span
+	keyPrefix, err := ElidedPrefix(newKey, s.conf.ElideMode)
+	if err != nil {
+		return err
+	}
+	if newKey.Compare(s.prevKey) < 0 || !bytes.Equal(keyPrefix, s.elidedPrefix) {
+		log.VEventf(
+			ctx, 1, "flushing backup file %s of size %d because key %s cannot append before %s",
+			s.outName, s.flushedSize, newKey, s.prevKey,
+		)
+		s.stats.oooFlushes++
+		if err := s.Flush(s.ctx); err != nil {
+			return err
+		}
+		// After a flush, we need to reset the span for the new key.
+		return s.Reset(ctx, lastSpan)
+	}
+	return nil
+}
+
+// maybeDoSizeFlushWithNextKey is checks if a flush is needed and info
+// about the next key to be written is provided. This is used to determine if
+// the next key can act as an exclusive end key for the last BackupManifest_File.
+// If a flush is performed, then a new span will be created with Reset to
+// contiguously extend the last span. This function should only be called after
+// s.midRow has been updated for the last key/span written.
+func (s *SSTSinkKeyWriter) maybeDoSizeFlushWithNextKey(
+	ctx context.Context, nextKey roachpb.Key,
+) error {
+	if len(s.flushedFiles) == 0 || !s.shouldDoSizeFlush() {
+		log.VEventf(ctx, 3, "continuing to write to backup file %s of size %d", s.outName, s.flushedSize)
+		return nil
+	}
+	// If the span of the next key to be written covers the same span as the last
+	// BackupManifest_File, then the next key can act as an exclusive end key for
+	// the last file's span. Additionally, the span of the nextKey will be updated
+	// so that it contiguously extends with the last span.
```

Comment:
this feels cleaner:
```
spanEndKey := &s.flushedFiles[len(s.flushedFiles)-1].EndKey
newSpan := := roachpb.Span{
		Key:    nextKey,
		EndKey: *spanEndKey,
	}
s.flushedFiles[len(s.flushedFiles)-1].EndKey = nextKey
err := s.doSizeFlush(ctx)....
```
We should double check that this doesn't escape to the heap. If it does, I think we'll need to create a scratch key on the key writer. Lets worry about this a little later.

Reply:
er, see comment about file entry based flushes.

Reply:
Maybe something like:
```golang
lastSpan := &s.flushedFiles[len(s.flushedFiles)-1].Span
newSpan := roachpb.Span{
		Key:    nextKey,
		EndKey: lastSpan.EndKey,
	}
lastSpan.EndKey = nextKey
err := s.doSizeFlush(ctx)
```

Dereferencing the `EndKey` doesn't seem to be too useful